contracts-periphery/src/UniswapWithdrawHook.sol

```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+pragma abicoder v2;
+
+import "openzeppelin-contracts/contracts/access/Ownable.sol";
+import "openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+import "src/interfaces/ISwapRouter.sol";
+
+contract UniswapWithdrawHook is Ownable {
+  using SafeERC20 for IERC20;
+
+  ISwapRouter internal immutable swapRouter;
+  address internal WETH9;
+
+  uint256 public feeBips;
+  address public feeCollector;
+  address payable public feeReceiver;
+
+  constructor(ISwapRouter _swapRouter, address _WETH9, uint256 _feeBips, address payable _feeReceiver) {
+  swapRouter = _swapRouter;
+  WETH9 = _WETH9;
+  feeBips = _feeBips;
+  feeReceiver = _feeReceiver;
+  }
+
+  function tokensWithdrawn(
+    uint256 _amount,
+    address _stealthAddr,
+    address _acceptor,
+    address _tokenAddr,
+    address _sponsor,
+    uint256 _sponsorFee,
+    bytes memory _data
+  ) external{
+
+    (bool _singleTrade, , ) = abi.decode(_data, (bool, address, ISwapRouter.ExactInputParams[]));
+    if(_singleTrade) {
+      _swapPartForEth(_amount, _stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, _data);
+    } else {
+      _swapPartForTokenPartForEth(_amount, _stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, _data);
+    }
+
+  }
+
+  function _swapPartForEth(
+    uint256 _amount,
+    address _stealthAddr,
+    address _acceptor,
+    address _tokenAddr,
+    address _sponsor,
+    uint256 _sponsorFee,
+    bytes memory _data) internal {
+    uint256 senderBalance = _amount;
+    ( , address _recipient, ISwapRouter.ExactInputParams[] memory params) = abi.decode(_data, (bool, address, ISwapRouter.ExactInputParams[]));
+    senderBalance -= params[0].amountIn;
+
+    bytes[] memory data = new bytes[](2);
+    data[0] = abi.encodeWithSelector(swapRouter.exactInput.selector, params[0]);
+    // Need to rethink usage of params[0].amountOutMinimum below
+    data[1] = abi.encodeWithSelector(swapRouter.unwrapWETH9WithFee.selector, params[0].amountOutMinimum, _recipient, feeBips, feeReceiver);
+
+    swapRouter.multicall(data);
+    if(senderBalance > 0) {
+      IERC20(_tokenAddr).safeTransferFrom(address(this), _recipient, senderBalance);
+    }
+  }
+
+  function _swapPartForTokenPartForEth(
```

Comment:
I think a good way to implement this is similar to the above comment, where the `data` includes:
- One encoded multicall to execute the first swap to ETH and send it to the recipient
- Some struct or signal that says "swap the full balance of token X, expect a minimum output of Y, and swap into token Z and send it to the recipient". This is required because the caller won't know the remaining token X balance prior to swapping

Reply:
Currently we use the multicall data that's passed in and send the leftover balance of tokens to the final recipient. https://github.com/ScopeLift/umbra-protocol/pull/341/commits/b40d4a8da174dc150d24e62c76a20189c2d8cfc2